File: 1626.best-team-with-no-conflicts.cs
/*
 * @lc app=leetcode id=1626 lang=csharp
 *
 * [1626] Best Team With No Conflicts
 */

// @lc code=start
public class maxSegmentTree
{
    public int[] tree;
    public maxSegmentTree(int size, int initialValue)
    {
        tree = new int[size];
        Array.Fill(tree, initialValue);
    }

    public void update(int i, int val)
    {
        i += tree.Length / 2;
        tree[i] = val;
        i = i / 2;
        while (i > 0)
        {
            tree[i] = Math.Max(tree[2*i], tree[2*i + 1]);
            i = i / 2;
        }
    }

    public int query(int start, int end)
    {
        start += tree.Length / 2;
        end += tree.Length / 2;
        int res = 0;
        while (start <= end)
        {
            if (start % 2 == 1)
            {
                res = Math.Max(res, tree[start]);
                start++;
            }
            if (end % 2 == 0)
            {
                res = Math.Max(res, tree[end]);
                end--;
            }
            start /= 2;
            end /= 2;
        }
        return res;
    }
}
public class Solution 
{
    public int BestTeamScore(int[] scores, int[] ages) 
    {
        maxSegmentTree dp = new maxSegmentTree(2*(1 + ages.Max()), 0);
        List<int[]> data = new List<int[]>();
        int result = 0;
        for (int i = 0; i < ages.Length; i++)
        {
            data.Add(new int[2] {scores[i], ages[i]});
        }
        data = data.OrderBy(x => x[0]).ThenBy(x => x[1]).ToList();

        foreach (int[] datum in data)
        {
            int newMaxTotalScoreCandidate = datum[0] + dp.query(0, datum[1]);
            result = Math.Max(result, newMaxTotalScoreCandidate);
            dp.update(datum[1], newMaxTotalScoreCandidate);
        }
        return result;
    }
}
// @lc code=end

File: 1626.best-team-with-no-conflicts.py
#
# @lc app=leetcode id=1626 lang=python3
#
# [1626] Best Team With No Conflicts
#

# @lc code=start
from typing import List

class MaxSegmentTree:
    def __init__(self, n, value) -> None:
        self.tree = [value]*n

    def update(self, i, val):
        i += len(self.tree) // 2
        self.tree[i] = val
        i //= 2
        while i > 0:
            self.tree[i] = max(self.tree[2*i], self.tree[2*i + 1])
            i //= 2
    
    def query(self, i, j):
        i += len(self.tree) // 2
        j += len(self.tree) // 2
        s = 0
        while i <= j:
            if i % 2 == 1:
                s = max(s, self.tree[i])
                i += 1
            if j % 2 == 0:
                s = max(s, self.tree[j])
                j -= 1
            i //= 2
            j //= 2
        return s 

class Solution:
    # We first group the data together in terms of [score, age]
    # Then sort them by score then by age
    # Then for each item, we keep adding them to a segment tree of size max(ages)
    # Because scores are nondecreasing, for each new item we ask
    # what is the maximum total score so far up to current age? Then add to it because the current score must be larger than or equal to the previous max score
  
    def bestTeamScore(self, scores: List[int], ages: List[int]) -> int:
        dp = MaxSegmentTree(2*(1 + max(ages)), 0)
        data = []
        result = 0
        for score, age in zip(scores, ages):
            data.append([score, age])
        data.sort()

        for score, age in data:
            maxTotalScoreSoFar = dp.query(0, age)
            newMaxTotalScoreCandidate = maxTotalScoreSoFar + score
            result = max(result, newMaxTotalScoreCandidate)
            dp.update(age, newMaxTotalScoreCandidate)
        return result


# @lc code=end

